libraries/AP_Airspeed/AP_Airspeed_AUAV.cpp

```diff
@@ -0,0 +1,246 @@
+
+#define MEASUREMENT_TIMEOUT_MS 200
+
+#include <AP_Common/AP_Common.h>
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+#include <AP_Math/AP_Math.h>
+#include <GCS_MAVLink/GCS.h>
+
+extern const AP_HAL::HAL &hal;
+
+
+AP_Airspeed_AUAV::AP_Airspeed_AUAV(AP_Airspeed &_frontend, uint8_t _instance, const float _range_inH2O) :
+    AP_Airspeed_Backend(_frontend, _instance),
+    range_inH2O(_range_inH2O)
+{
+}
+
+// probe for a sensor
+bool AP_Airspeed_AUAV::probe(uint8_t bus, uint8_t address)
+{
+    _dev = hal.i2c_mgr->get_device(bus, address);
+    if (!_dev) {
+        return false;
+    }
+
+    WITH_SEMAPHORE(_dev->get_semaphore());
+
+    _dev->set_retries(10);
+    _measure();
+    hal.scheduler->delay(10);
+    _collect();
+
+    return last_sample_time_ms != 0;
+
+}
+
+// probe and initialise the sensor
+bool AP_Airspeed_AUAV::init()
+{
+    if (bus_is_configured()) {
+        // the user has configured a specific bus
+        if (probe(get_bus(), AUAV_AIRSPEED_I2C_ADDR)) {
+            goto found_sensor;
+        }
+    } else {
+        // if bus is not configured then fall back to the old
+        // behaviour of probing all buses, external first
+        FOREACH_I2C_EXTERNAL(bus) {
+            if (probe(bus, AUAV_AIRSPEED_I2C_ADDR)) {
+                goto found_sensor;
+            }
+        }
+        FOREACH_I2C_INTERNAL(bus) {
+            if (probe(bus, AUAV_AIRSPEED_I2C_ADDR)) {
+                goto found_sensor;
+            }
+        }
+    }
+
+    GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "AUAV AIRSPEED[%u]: no sensor found", get_instance());
+    return false;
+
+found_sensor:
+    _dev->set_device_type(uint8_t(DevType::AUAV));
+    set_bus_id(_dev->get_bus_id());
+
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "AUAV AIRSPEED[%u]: Found bus %u addr 0x%02x", get_instance(), _dev->bus_num(), _dev->get_bus_address());
+
+    // drop to 2 retries for runtime
+    _read_coefficients();
+    _dev->set_retries(2);
+    _dev->register_periodic_callback(20000,
+                                     FUNCTOR_BIND_MEMBER(&AP_Airspeed_AUAV::_timer, void));
+    return true;
+}
+
+// start a measurement
+void AP_Airspeed_AUAV::_measure()
+{
+    measurement_started_ms = 0;
+    uint8_t cmd = 0xAD; // start average8, max 16.2ms measurement time
+    if (_dev->transfer(&cmd, 1, nullptr, 0)) {
+        measurement_started_ms = AP_HAL::millis();
+    }
+}
+
+// read the values from the sensor
+void AP_Airspeed_AUAV::_collect()
+{
+    measurement_started_ms = 0; // It should always get reset by _measure. This is a safety to handle failures of i2c bus
+    uint8_t inbuf[7];
+    if (!_dev->read((uint8_t *)&inbuf, sizeof(inbuf))) {
+        return;
+    }
+    const int32_t Tref_Counts = 7576807; // temperature counts at 25C
+    const float TC50Scale = 100.0f * 100.0f * 167772.2f; // scale TC50 to 1.0% FS0
+    float AP3, BP2, CP, Corr, Pcorr, Pdiff, TC50, Pnfso, Tcorr, PCorrt;
+    int32_t iPraw, Tdiff, iTemp;
+    uint32_t PComp;
```

Comment:
Would be nice to initialise these as we declare them.

Reply:
done